File: BAEKJOON/2108 - 통계학/통계학.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <map>
using namespace std;

int main() {
    // 수의 개수 N 입력
    int N;
    cin >> N;

    // 입력된 수들을 저장할 벡터와 빈도를 저장할 map 선언
    vector<int> arr(N);
    map<int, int> frequency;

    // 수 입력 및 빈도 계산
    for (int i = 0; i < N; ++i) {
        cin >> arr[i];
        frequency[arr[i]]++;
    }

    // 벡터를 오름차순으로 정렬
    sort(arr.begin(), arr.end());

    // 산술평균 계산 및 출력
    double sum = 0;
    for (int num : arr) sum += num;
    // round 함수는 소수점 첫째 자리에서 반올림된 값을 반환
    int mean = round(sum / N);
    cout << mean << '\n';

    // 중앙값은 정렬된 벡터에서 가운데 값
    int median = arr[N / 2];
    cout << median << '\n';

    // 최빈값 찾기
    vector<pair<int, int>> freqVec(frequency.begin(), frequency.end());
    // 빈도수가 높은 순서로 정렬, 빈도가 같으면 숫자가 작은 순서로 정렬
    sort(freqVec.begin(), freqVec.end(), [](const pair<int, int>& a, const pair<int, int>& b) {
        if (a.second == b.second) return a.first < b.first;
        return a.second > b.second;
    });

    // 최빈값이 여러 개일 경우 두 번째로 작은 값 출력
    int mode = (freqVec.size() > 1 && freqVec[0].second == freqVec[1].second) 
               ? freqVec[1].first 
               : freqVec[0].first;
    cout << mode << '\n';

    // 범위는 정렬된 벡터에서 최댓값과 최솟값의 차이
    int range = arr.back() - arr.front();
    cout << range << '\n';

    return 0;
}

File: BAEKJOON/2108 - 통계학/통계학.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main()
    {
        // 수의 개수 N 입력
        int N = int.Parse(Console.ReadLine());

        // 입력된 수들을 저장할 리스트와 빈도수를 저장할 딕셔너리 선언
        List<int> arr = new List<int>();
        Dictionary<int, int> frequency = new Dictionary<int, int>();

        // 수 입력 및 빈도 계산
        for (int i = 0; i < N; i++)
        {
            // 수 입력받기
            int num = int.Parse(Console.ReadLine());
            arr.Add(num);

            // 빈도수를 딕셔너리에 저장 (기존에 키가 있으면 증가, 없으면 추가)
            if (frequency.ContainsKey(num))
                frequency[num]++;
            else
                frequency[num] = 1;
        }

        // 리스트를 오름차순으로 정렬
        arr.Sort();

        // 산술평균 계산 및 출력
        double sum = arr.Sum(); // 리스트의 합을 구함
        int mean = (int)Math.Round(sum / N); // 평균을 구하고 반올림 후 정수형으로 변환
        Console.WriteLine(mean);

        // 중앙값 계산 (정렬된 리스트의 중앙값)
        int median = arr[N / 2];
        Console.WriteLine(median);

        // 빈도수 딕셔너리를 리스트로 변환하여 빈도수 기준으로 정렬
        var sortedFrequency = frequency.OrderByDescending(x => x.Value).ThenBy(x => x.Key).ToList();

        // 최빈값이 여러 개일 경우 두 번째로 작은 값을 선택
        int mode = (sortedFrequency.Count > 1 && sortedFrequency[0].Value == sortedFrequency[1].Value)
            ? sortedFrequency[1].Key
            : sortedFrequency[0].Key;
        Console.WriteLine(mode);

        // 범위 계산 (정렬된 리스트의 최댓값 - 최솟값)
        int range = arr.Last() - arr.First();
        Console.WriteLine(range);
    }
}